src/Multipart/AbstractUploader.php

```diff
@@ -0,0 +1,455 @@
+<?php
+namespace Aws\Multipart;
+
+use Aws\AwsClientInterface as Client;
+use Aws\CommandInterface;
+use Aws\CommandPool;
+use Aws\Exception\AwsException;
+use Aws\Exception\MultipartUploadException;
+use Aws\Result;
+use Aws\ResultInterface;
+use GuzzleHttp\Promise;
+use GuzzleHttp\Promise\PromiseInterface;
+use GuzzleHttp\Psr7;
+use InvalidArgumentException as IAE;
+use Psr\Http\Message\RequestInterface;
+use Psr\Http\Message\StreamableInterface as Stream;
+
+/**
+ * Encapsulates the execution of a multipart upload to S3 or Glacier.
+ */
+abstract class AbstractUploader
```

Comment:
What about making this implement PromisorInterface so that calling `promise()` will start an async upload, and calling cancel on the promise will abort the upload? That could possibly remove the need for an uploadAsync method.